typescript/sdk/src/token/EvmERC20WarpRouteReader.ts

```diff
@@ -112,16 +160,50 @@ export class EvmERC20WarpRouteReader {
    *
    * @param tokenAddress - The address of the token.
    * @returns A partial ERC20 metadata object containing the token name, symbol, total supply, and decimals.
+   * Throws if unsupported token type
    */
-  async fetchTokenMetadata(tokenAddress: Address): Promise<TokenMetadata> {
-    const erc20 = ERC20__factory.connect(tokenAddress, this.provider);
-    const [name, symbol, totalSupply, decimals] = await Promise.all([
+  async fetchTokenMetadata(
+    type: TokenType,
+    tokenAddress: Address,
+  ): Promise<TokenMetadata & { token?: string }> {
+    if (type === TokenType.collateral || type === TokenType.collateralVault) {
+      const erc20 = HypERC20Collateral__factory.connect(
+        tokenAddress,
+        this.provider,
+      );
+      const token = await erc20.wrappedToken();
+      const { name, symbol, decimals, totalSupply } =
+        await this.fetchERC20Metadata(token);
+
+      return { name, symbol, decimals, totalSupply, token };
+    } else if (type === TokenType.synthetic) {
+      return this.fetchERC20Metadata(tokenAddress);
+    } else if (type === TokenType.native) {
+      const chainMetadata = this.multiProvider.getChainMetadata(this.chain);
+      if (chainMetadata.nativeToken) {
+        const { name, symbol, decimals } = chainMetadata.nativeToken;
+        return { name, symbol, decimals, totalSupply: 0 };
+      } else {
+        throw new Error(
+          `Warp route config specifies native token but chain metadata for ${this.chain} does not provide native token details`,
+        );
+      }
+    } else {
+      throw new Error(
+        `Unsupported token type ${type} when fetching token metadata`,
+      );
+    }
+  }
+
+  async fetchERC20Metadata(tokenAddress: Address): Promise<TokenMetadata> {
```

Comment:
I think we have this behavior in a few different places, might be nice to consolidate